File: dotnet/Multiplayer/Rooms/Updates/Updates.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Michitai.Multiplayer.Rooms.Updates
{
    /// <summary>
    /// Provides static methods for managing room updates.
    /// Handles sending updates to players and polling for received updates.
    /// </summary>
    public class Updates
    {
        /// <summary>
        /// Sends updates to specific players in the room.
        /// </summary>
        /// <typeparam name="T">The type of update data.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="request">The update request containing targets and data.</param>
        /// <param name="ct">Cancellation token.</param>
        /// <returns>Task containing the update players response.</returns>
        public static Task<UpdatePlayersResponse> UpdatePlayersAsync<T>(Client client, string playerToken, UpdatePlayers<T> request, CancellationToken ct = default) where T : class, new()
            => client.Send<UpdatePlayersResponse>(HttpMethod.Post, client.Url(Endpoints.GameRoomUpdates, $"&player_token={playerToken}"),
                new UpdatePlayersRequest<T>(request.Target_players, request.Type, request.Data, request.Target_players_ids), ct);

        /// <summary>
        /// Polls for updates that were sent to the current player.
        /// </summary>
        /// <typeparam name="T">The type of update data.</typeparam>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's authentication token.</param>
        /// <param name="request">The poll request containing filters.</param>
        /// <param name="ct">Cancellation token.</param>
        /// <returns>Task containing the poll updates response with received updates.</returns>
        public static Task<PollUpdatesResponse<T>> PollUpdatesAsync<T>(Client client, string playerToken, PollUpdates request,
            CancellationToken ct = default) where T : class, new()
                => client.Send<PollUpdatesResponse<T>>(HttpMethod.Post, client.Url(Endpoints.GameRoomUpdatesPoll, $"&player_token={playerToken}"),
                    new PollUpdatesRequest(request.From_players, request.From_players_ids, request.Last_update), ct);
    }
}

File: python/multiplayer/rooms/updates/updates.py
"""Room updates: send updates to players and poll received updates."""

from __future__ import annotations

import dataclasses
from typing import Any, List, Optional, Type, TypeVar

from ..._serialization import from_dict
from ...client import Client
from ..enums import RoomTargetPlayers
from .models import PollUpdatesResponse, UpdatePlayersResponse

T = TypeVar("T")

# Endpoints
GAME_ROOM_UPDATES = "game_room.php/updates"
GAME_ROOM_UPDATES_POLL = "game_room.php/updates/poll"


class Updates:
    """Send and receive room updates (polling-based state sync)."""

    def __init__(self, client: Client) -> None:
        self._client = client

    def send(
        self,
        player_token: str,
        type: str,
        data: Any = None,
        target_players: RoomTargetPlayers = RoomTargetPlayers.ALL,
        target_players_ids: Optional[List[int]] = None,
    ) -> UpdatePlayersResponse:
        """Send an update to specific players in the room.

        :param type: The update type (e.g. ``"game_start"``).
        :param data: The update payload.
        :param target_players: Which players to target.
        :param target_players_ids: Player IDs when targeting SPECIFIC.
        """
        return self._client.send(
            "POST",
            self._client.player_url(GAME_ROOM_UPDATES, player_token),
            {
                "target_players": target_players.value,
                "target_players_ids": target_players_ids,
                "type": type,
                "data": data,
            },
            UpdatePlayersResponse,
        )

    def poll(
        self,
        player_token: str,
        from_players: RoomTargetPlayers = RoomTargetPlayers.HOST,
        from_players_ids: Optional[List[int]] = None,
        last_update: Optional[str] = None,
        data_cls: Optional[Type[T]] = None,
    ) -> PollUpdatesResponse:
        """Poll updates that were sent to the current player.

        :param from_players: Which players to receive updates from.
        :param from_players_ids: Player IDs when ``from_players`` is SPECIFIC.
        :param last_update: Only receive updates after this update ID.
        :param data_cls: Optional dataclass type for each update's ``data``.
        """
        response = self._client.send(
            "POST",
            self._client.player_url(GAME_ROOM_UPDATES_POLL, player_token),
            {
                "from_players": from_players.value,
                "from_players_ids": from_players_ids,
                "last_update": last_update,
            },
            PollUpdatesResponse,
        )
        if data_cls is not None:
            for update in response.updates:
                if dataclasses.is_dataclass(data_cls) and isinstance(update.data, dict):
                    update.data = from_dict(data_cls, update.data)
        return response
